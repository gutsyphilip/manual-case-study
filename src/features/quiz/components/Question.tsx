import React, { useEffect } from "react";
import styles from "./quiz.module.scss";
import cx from "classnames";

interface Props {
  data: {
    question: string;
    type: string;
    options: { [key: string]: string | boolean }[];
  };
  number: number;
  activeQuestion: number;
  // TODO: Resolve ref type issue
  activeRef?: any;
  userResponses: { [key: string]: string | boolean };
  handleSetResponse: (
    question: string,
    response: string | boolean,
    questionNumber: number
  ) => void;
}

const Question = ({
  data: { options, question, type },
  number,
  activeQuestion,
  handleSetResponse,
  userResponses,
  activeRef,
}: Props) => {
  const renderOptions = (): React.ReactNode => {
    return options.map((option, index) => {
      const label: string = Object.keys(option)[0];
      return (
        <div
          className={cx(styles.quizModal_question_options_item, {
            [styles.quizModal_question_options_item__selected]:
              userResponses[question] === options[index][label],
          })}
          data-testid="option"
          key={`${label}`}
          onClick={() => {
            handleSetResponse(question, options[index][label], number);
          }}
        >
          <div
            className={styles.quizModal_question_options_item_label}
            dangerouslySetInnerHTML={{ __html: label }}
          ></div>
          <div className={styles.quizModal_question_options_item_text}>
            {options[index][label]}
          </div>
        </div>
      );
    });
  };

  useEffect(() => {
    if (activeQuestion > 1) {
      activeRef?.current?.scrollIntoView({
        behavior: "smooth",
        block: "end",
        inline: "nearest",
      });
    }
  }, [activeRef, activeQuestion]);

  return (
    <div
      className={cx(styles.quizModal_question, {})}
      ref={activeQuestion === number ? activeRef : null}
    >
      <h4
        className={styles.quizModal_question_label}
      >{`Question ${number}`}</h4>
      <h1 className={styles.quizModal_question_title}>{question}</h1>

      <div
        className={styles.quizModal_question_options}
        data-testid="options-list"
      >
        {renderOptions()}
      </div>
    </div>
  );
};

export default Question;
