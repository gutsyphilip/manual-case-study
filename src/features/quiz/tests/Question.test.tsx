import React, { useRef } from "react";
import { render } from "@testing-library/react";
import Question from "../components/Question";

const QuestionProps = {
  data: {
    question: "Which image best matches your hair loss?",
    type: "ChoiceType",
    options: [
      {
        '<img alt="Temples"     src="https://s3-eu-west-1.amazonaws.com/manualco/questions/temples-hairloss.png" srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/temples-hairloss%402x.png 2x" />':
          "Temples",
      },
      {
        '<img alt="Temples & Crown" src="https://s3-eu-west-1.amazonaws.com/manualco/questions/templescrown-hairloss.png" srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/templescrown-hairloss%402 x.png 2x"/>':
          "Temples & Crown",
      },
      {
        '<img alt="Patchy" src="https://s3-eu-west-1.amazonaws.com/manualco/questions/patchy-hairloss.png" srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/patchy-hairloss%402x.png 2x"/>':
          "Patchy",
      },
      {
        '<img alt="Moderate" src="https://s3-eu-west-1.amazonaws.com/manualco/questions/moderate-hairloss.png" srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/moderate-hairloss%402x.pn g 2x" />':
          "Moderate",
      },
      {
        '<img alt="Extensive" src="https://s3-eu-west-1.amazonaws.com/manualco/questions/extensive-hairloss.png" srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/extensive-hairloss%402x.pn g 2x"/>':
          "Extensive",
      },
      {
        '<img alt="Complete" src="https://s3-eu-west-1.amazonaws.com/manualco/questions/complete-hairloss.png"     srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/complete-hairloss%402x.pn g 2x" />':
          "Complete",
      },
    ],
  },
  number: 1,
  activeQuestion: 1,
  handleSetResponse: jest.fn(() => {}),
  userResponses: {},
};

test("question is properly displayed to user", () => {
  const { getByLabelText } = render(<Question {...QuestionProps} />);
});
