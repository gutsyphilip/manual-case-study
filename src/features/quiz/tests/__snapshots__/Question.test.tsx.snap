// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options are properly displayed to user 1`] = `
<div
  class="quizModal_question_options"
  data-testid="options-list"
>
  <div
    class="quizModal_question_options_item"
    data-testid="option"
  >
    <div
      class="quizModal_question_options_item_label"
    >
      <img
        alt="Temples"
        src="https://s3-eu-west-1.amazonaws.com/manualco/questions/temples-hairloss.png"
        srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/temples-hairloss%402x.png 2x"
      />
    </div>
    <div
      class="quizModal_question_options_item_text"
    >
      Temples
    </div>
  </div>
  <div
    class="quizModal_question_options_item"
    data-testid="option"
  >
    <div
      class="quizModal_question_options_item_label"
    >
      <img
        alt="Temples & Crown"
        src="https://s3-eu-west-1.amazonaws.com/manualco/questions/templescrown-hairloss.png"
        srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/templescrown-hairloss%402 x.png 2x"
      />
    </div>
    <div
      class="quizModal_question_options_item_text"
    >
      Temples & Crown
    </div>
  </div>
  <div
    class="quizModal_question_options_item"
    data-testid="option"
  >
    <div
      class="quizModal_question_options_item_label"
    >
      <img
        alt="Patchy"
        src="https://s3-eu-west-1.amazonaws.com/manualco/questions/patchy-hairloss.png"
        srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/patchy-hairloss%402x.png 2x"
      />
    </div>
    <div
      class="quizModal_question_options_item_text"
    >
      Patchy
    </div>
  </div>
  <div
    class="quizModal_question_options_item"
    data-testid="option"
  >
    <div
      class="quizModal_question_options_item_label"
    >
      <img
        alt="Moderate"
        src="https://s3-eu-west-1.amazonaws.com/manualco/questions/moderate-hairloss.png"
        srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/moderate-hairloss%402x.pn g 2x"
      />
    </div>
    <div
      class="quizModal_question_options_item_text"
    >
      Moderate
    </div>
  </div>
  <div
    class="quizModal_question_options_item"
    data-testid="option"
  >
    <div
      class="quizModal_question_options_item_label"
    >
      <img
        alt="Extensive"
        src="https://s3-eu-west-1.amazonaws.com/manualco/questions/extensive-hairloss.png"
        srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/extensive-hairloss%402x.pn g 2x"
      />
    </div>
    <div
      class="quizModal_question_options_item_text"
    >
      Extensive
    </div>
  </div>
  <div
    class="quizModal_question_options_item"
    data-testid="option"
  >
    <div
      class="quizModal_question_options_item_label"
    >
      <img
        alt="Complete"
        src="https://s3-eu-west-1.amazonaws.com/manualco/questions/complete-hairloss.png"
        srcset="https://s3-eu-west-1.amazonaws.com/manualco/questions/complete-hairloss%402x.pn g 2x"
      />
    </div>
    <div
      class="quizModal_question_options_item_text"
    >
      Complete
    </div>
  </div>
</div>
`;

exports[`question is properly displayed to user 1`] = `
<h1
  class="quizModal_question_title"
>
  Which image best matches your hair loss?
</h1>
`;
